provider/provider.go

```diff
@@ -585,21 +584,43 @@ func hashContext(dockerContextPath string, dockerfile string) (string, error) {
 	return accumulator.hexSumContext(), nil
 }
 
-func getIgnore(dockerIgnorePath string) ([]string, error) {
-	var ignorePatterns []string
-	dockerIgnore, err := os.ReadFile(dockerIgnorePath)
-	if err != nil {
-		if os.IsNotExist(err) {
-			// .dockerignore files are optional
-			return ignorePatterns, nil
+// getIgnorePatterns returns all patterns to ignore when constructing a build
+// context for the given Dockerfile, if any such patterns exists.
+//
+// Precedence is given to Dockerfile-specific ignore-files as per
+// https://docs.docker.com/build/building/context/#filename-and-location.
+func getIgnorePatterns(fs afero.Fs, dockerfilePath, contextRoot string) ([]string, error) {
+	paths := []string{}
+
+	// If we have a non-standard Dockerfile, first attempt to use its specific
+	// ignore-file, if one exists.
+	if filepath.Base(dockerfilePath) != defaultDockerfile {
+		paths = append(paths, dockerfilePath+".dockerignore")
+	}
+
+	// Always fall back to the ignore-file at the root of our build context.
+	paths = append(paths, filepath.Join(contextRoot, ".dockerignore"))
+
+	// Attempt to parse our candidate ignore-files, skipping any that don't
+	// exist.
+	for _, p := range paths {
+		f, err := fs.Open(p)
+		if errors.Is(err, afero.ErrFileNotFound) {
+			continue
 		}
-		return ignorePatterns, fmt.Errorf("unable to read %s file: %w", dockerIgnorePath, err)
-	}
-	ignorePatterns, err = dockerignore.ReadAll(bytes.NewReader(dockerIgnore))
-	if err != nil {
-		return ignorePatterns, fmt.Errorf("unable to parse %s file: %w", ".dockerignore", err)
+		if err != nil {
+			return nil, fmt.Errorf("reading %q: %w", p, err)
+		}
+		defer f.Close()
+
+		ignorePatterns, err := dockerignore.ReadAll(f)
```

Comment:
I presume we return on the first found `.dockerignore` file because we're relying on the order in which the particular `.dockerignore`s get appended to `paths` (specific first, root-level second) - is there a way to make that precedence more explicit?

Reply:
Yes, exactly. I did some spelunking and confirmed we should always check for the present of `<Dockerfile>.dockerignore` regardless of the file's name. So that makes this logic a little more explicit -- there are always 2 things to check and in a particular order.